ts/components/conversation/message/message-content/MessageReactBar.tsx

```diff
@@ -51,45 +51,132 @@ const ReactButton = styled.span`
+  const expirationTimestamp = useIsRenderedExpiresInItem(messageId);
 
-  useMount(() => {
-    const reactions = new RecentReactions(getRecentReactions());
-    if (reactions && !isEqual(reactions, recentReactions)) {
-      setRecentReactions(reactions);
-    }
-  });
+  return (
+    <StyledContainer expirationTimestamp={expirationTimestamp}>
+      <StyledMessageReactBar>
+        {recentReactions &&
+          recentReactions.map(emoji => (
+            <ReactButton
+              key={emoji}
+              role={'img'}
+              aria-label={
+                nativeEmojiData?.ariaLabels ? nativeEmojiData.ariaLabels[emoji] : undefined
+              }
+              onClick={() => {
+                action(emoji);
+              }}
+            >
+              {emoji}
+            </ReactButton>
+          ))}
+        <SessionIconButton
+          iconColor={'var(--emoji-reaction-bar-icon-color)'}
+          iconPadding={'8px'}
+          iconSize={'huge'}
+          iconType="plusThin"
+          backgroundColor={'var(--emoji-reaction-bar-icon-background-color)'}
+          borderRadius="300px"
+          onClick={additionalAction}
+        />
+      </StyledMessageReactBar>
+      <ExpiresInItem expirationTimestamp={expirationTimestamp} />
+    </StyledContainer>
+  );
+};
+
+function formatExpiry({ expirationTimestamp }: { expirationTimestamp: number }) {
+  const diffMs = expirationTimestamp - Date.now();
+  const diff = moment(diffMs).utc();
+
+  if (diffMs <= 0) {
+    return `0s`;
+  }
+
+  const prefix = 'Message will expire in';
+
+  if (diff.isBefore(moment.utc(0).add(1, 'minute'))) {
+    return `${prefix} ${diff.seconds()}s`;
+  }
+
+  if (diff.isBefore(moment.utc(0).add(1, 'hour'))) {
+    const extraUnit = diff.seconds() ? ` ${diff.seconds()}s` : '';
+    return `${prefix} ${diff.minutes()}m${extraUnit}`;
+  }
+
+  if (diff.isBefore(moment.utc(0).add(1, 'day'))) {
+    const extraUnit = diff.minutes() ? ` ${diff.minutes()}m` : '';
+    return `${prefix} ${diff.hours()}h${extraUnit}`;
+  }
+
+  if (diff.isBefore(moment.utc(0).add(7, 'day'))) {
+    const extraUnit = diff.hours() ? ` ${diff.hours()}h` : '';
+    return `${prefix} ${diff.dayOfYear() - 1}d${extraUnit}`;
+  }
+
+  if (diff.isBefore(moment.utc(0).add(31, 'day'))) {
+    const days = diff.dayOfYear() - 1;
+    const weeks = Math.floor(days / 7);
+    const daysLeft = days % 7;
+    const extraUnit = daysLeft ? ` ${daysLeft}d` : '';
+    return `${prefix} ${weeks}w${extraUnit}`;
+  }
+
+  return '...';
+}
+
+const StyledExpiresIn = styled.div`
+  border-radius: 8px;
+  padding: 10px;
+  white-space: nowrap;
+  color: var(--text-primary-color);
+  size: var(--font-size-sm);
+  background-color: var(--context-menu-background-color);
+  box-shadow: 0px 0px 9px 0px var(--context-menu-shadow-color);
+  margin-top: 7px;
+  display: flex;
+  align-items: center;
+  min-width: 0;
+`;
+
+function useIsRenderedExpiresInItem(messageId: string) {
```

Comment:
maybe move into a hooks.ts file in the disappearing_messages folder? Again trying to keep feature code together.